File: backend/src/test/java/lan/dk/podcastserver/manager/worker/rss/RSSUpdaterTest.java
package lan.dk.podcastserver.manager.worker.rss;

import com.github.davinkevin.podcastserver.service.ImageService;
import com.github.davinkevin.podcastserver.service.SignatureService;
import io.vavr.collection.Set;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.manager.worker.Type;
import com.github.davinkevin.podcastserver.service.JdomService;
import com.github.davinkevin.podcastserver.IOUtils;
import org.jdom2.JDOMException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.runners.MockitoJUnitRunner;

import java.io.IOException;
import java.net.URISyntaxException;

import static io.vavr.API.None;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.AdditionalMatchers.not;
import static org.mockito.Mockito.*;

/**
 * Created by kevin on 28/06/15 for Podcast Server
 */
@RunWith(MockitoJUnitRunner.class)
public class RSSUpdaterTest {

    private static final String PODCAST_APPLOAD_URL = "/remote/podcast/rss/rss.appload.xml";
    private static final String MOCK_URL = "http://mockUrl.com/";
    private Podcast rssAppload;

    private @Mock SignatureService signatureService;
    private @Mock JdomService jdomService;
    private @Mock ImageService imageService;
    private @InjectMocks
    RSSUpdater rssUpdater;

    @Before
    public void beforeEach() throws JDOMException, IOException, URISyntaxException {
        rssAppload = Podcast.builder().url(MOCK_URL).build();

        when(jdomService.parse(eq(MOCK_URL))).thenReturn(IOUtils.fileAsXml(PODCAST_APPLOAD_URL));
        when(jdomService.parse(not(eq(MOCK_URL)))).thenReturn(None());
    }

    @Test
    public void should_get_items() throws JDOMException, IOException {
        /* Given */
        when(imageService.getCoverFromURL(anyString())).then(this::createCover);

        /* When */
        Set<Item> items = rssUpdater.getItems(rssAppload);

        /* Then */
        verify(jdomService, times(1)).parse(eq(MOCK_URL));
        verify(imageService, times(2)).getCoverFromURL(anyString());
        assertThat(items).hasSize(217)
                .anySatisfy(i -> assertThat(i.getCover()).isNotEqualTo(Cover.DEFAULT_COVER));
    }

    @Test
    public void should_return_null_if_not_updatable_podcast() {
        /* Given */ Podcast podcastNotUpdatable = new Podcast().setUrl("http://notUpdatable.com");
        /* When */  Set<Item> items = rssUpdater.getItems(podcastNotUpdatable);
        /* Then */  assertThat(items).isEmpty();
    }

    @Test
    public void should_return_an_empty_set() {
        /* Given */ Podcast podcast = Podcast.builder().url("http://foo.bar.fake.url/").build();
        /* When */  Set<Item> items = rssUpdater.getItems(podcast);
        /* Then */  assertThat(items).isEmpty();
    }

    @Test
    public void should_call_signature_from_url() {
        /* When */ rssUpdater.signatureOf(rssAppload);
        /* Then */ verify(signatureService, times(1)).fromUrl(eq(MOCK_URL));
    }

    @Test
    public void should_return_his_type() {
        Type type = rssUpdater.type();
        assertThat(type.key()).isEqualTo("RSS");
        assertThat(type.name()).isEqualTo("RSS");
    }

    private Cover createCover(InvocationOnMock i) {
        return Cover.builder().url(i.getArgument(0)).build();
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/manager/worker/rss/RSSUpdaterTest.kt
package com.github.davinkevin.podcastserver.manager.worker.rss

import arrow.core.None
import com.github.davinkevin.podcastserver.IOUtils.fileAsXml
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.service.JdomService
import com.github.davinkevin.podcastserver.service.SignatureService
import com.github.davinkevin.podcastserver.utils.toVΛVΓ
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.whenever
import com.github.davinkevin.podcastserver.entity.Cover
import com.github.davinkevin.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.manager.worker.PodcastToUpdate
import com.github.davinkevin.podcastserver.service.CoverInformation
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import java.net.URI
import java.time.ZoneOffset
import java.util.*

/**
 * Created by kevin on 28/06/15 for Podcast Server
 */
@ExtendWith(MockitoExtension::class)
class RSSUpdaterTest {

    @Mock lateinit var signatureService: SignatureService
    @Mock lateinit var jdomService: JdomService
    @Mock lateinit var imageService: ImageService
    @InjectMocks lateinit var updater: RSSUpdater

    private val rssAppload: PodcastToUpdate = PodcastToUpdate(
            url = URI("http://mockUrl.com/"),
            id = UUID.randomUUID(),
            signature = "noSign"
    )

    @Test
    fun `should get items`() {
        /* Given */
        val podcastUrl = "http://mockUrl.com/"
        whenever(jdomService.parse(podcastUrl)).thenReturn(fileAsXml("/remote/podcast/rss/rss.appload.xml"))
        whenever(imageService.fetchCoverInformation(any())).then { CoverInformation (
                url = URI(it.getArgument(0)),
                height = 200,
                width = 200
        ) }

        /* When */
        val items = updater.findItems(rssAppload)

        /* Then */
        assertThat(items).hasSize(217)
        assertThat(items.filter { it.cover == null }).hasSize(215)
        assertThat(items.filter { it.pubDate == null }).hasSize(1)

        assertThat(items.filter { it.pubDate?.offset == ZoneOffset.ofHours(6) }).hasSize(1)
        assertThat(items.filter { it.pubDate?.offset == ZoneOffset.ofHours(8) }).hasSize(2)
        assertThat(items.filter { it.pubDate?.offset == ZoneOffset.ofHours(9) }).hasSize(1)

        verify(jdomService, times(1)).parse(podcastUrl)
        verify(imageService, times(2)).fetchCoverInformation(any())
    }

    @Test
    fun `should return null if not updatable podcast`() {
        /* Given */
        val podcastNotUpdatable = rssAppload.copy( url = URI("http://notUpdatable.com") )
        whenever(jdomService.parse(any())).thenReturn(None.toVΛVΓ())

        /* When */
        val items = updater.findItems(podcastNotUpdatable)

        /* Then */
        assertThat(items).isEmpty()
    }

    @Test
    fun `should call signature from url`() {
        /* When */
        updater.signatureOf(rssAppload.url)

        /* Then */
        verify(signatureService, times(1)).fromUrl("http://mockUrl.com/")
    }

    @Test
    fun `should return its type`() {
        /* Given */
        /* When */
        val type = updater.type()
        /* Then */
        assertThat(type.key).isEqualTo("RSS")
        assertThat(type.name).isEqualTo("RSS")
    }

    @Test
    fun `should be compatible`() {
        /* Given */
        val url = "https://foo.bar.com"
        /* When */
        val compatibility = updater.compatibility(url)
        /* Then */
        assertThat(compatibility).isEqualTo(Integer.MAX_VALUE-1)
    }

    @Test
    fun `should not be compatible`() {
        /* Given */
        val url = "grpc://foo.bar.com"
        /* When */
        val compatibility = updater.compatibility(url)
        /* Then */
        assertThat(compatibility).isEqualTo(Integer.MAX_VALUE)
    }
}
